### DISCUSSION TITLE
Why is NextUInt64 not a public method

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/runtime#73266] [General] [answered]
https://github.com/dotnet/runtime/blob/508fef51e841aa16ffed1aae32bf4793a2cea363/src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs

In the source code the ability to produce NextUInt64 is clearly there but not available for System.Random so I suggest in one way or another allow me to produce UInt64 with System.Random instead of having to use stack allocated bytes as span to do so

**Comment by huoyaoyuan** [comment, marked as answer]
Because there's no `NextUInt32` either. Just cast the result of `NextInt64`.

**Reply by Xyncgas** [reply]
Oh yea I... I am sorry

**Reply by VesperGarment** [reply]
That doesn't work quite properly. NextInt64 only returns _positive_ integers - 63 bits of random. If you actually want all 64 bits random I think you have to do a little dance getting 4 bytes and BitConverter-ing it._

**Reply by elgonzo** [reply]
@VesperGarment that's not quite accurate. The `NextInt64(Int64, Int64)` overload can return negative values if you set the value range accordingly. For example, `NextInt64(long.MinValue, long.MaxValue)` should give you a value from almost the entire 64-bit value range. _Almost_ only because it won't ever return the `long.MaxValue` value.

**Comment by tannergooding** [comment]
Please feel free to open an API proposal request `UInt32` and `UInt64` APIs: https://github.com/dotnet/runtime/issues/new?assignees=&labels=api-suggestion&template=02_api_proposal.yml&title=%5BAPI+Proposal%5D%3A+

Our API Review Process is detailed here: https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md

While casting the result of `NextInt64` to `UInt64` will indeed work for some scenarios, it also makes the code significantly harder to read, understand, and maintain.

Given that integers in .NET are two's complement, getting values above `long.MaxValue` is also "difficult", because `long.MaxValue + 1` is `long.MinValue` and `-1` is `ulong.MaxValue`; so providing this functionality would allow valid unsigned ranges to be more easily returned.

**Reply by tannergooding** [reply]
Notably, we also use such APIs or similar APIs internally for our own tests and other scenarios. So it is completely within reason that other developers would benefit as well.
